File: src/flet_extension/flet_extension.py
from dataclasses import field
from typing import Optional, Any, Callable, Union, List

import flet as ft
from flet import AnimationValue, Animation, AnimationCurve, Duration, DurationValue

__all__ = ["FletExtension"]


@ft.control("FletExtension")
class FletExtension(ft.ConstrainedControl):
    """
    A comprehensive Flet extension control demonstrating advanced features.
    
    This control serves as a template for creating custom Flet extensions with:
    - Rich visual customization options
    - Animation support with multiple types and curves
    - Interactive event handling (click, hover)
    - Dynamic content updates
    - Professional styling capabilities
    
    Suitable for developers of all levels as a learning reference and starting point
    for building custom Flet controls.
    
    Example:
        ```python
        import flet as ft
        from flet_extension import FletExtension
        
        def main(page: ft.Page):
            ext = FletExtension(
                src="Hello, Flet!",
                title="My Extension",
                subtitle="Custom control demo",
                background_color="#e3f2fd",
                text_color="#1976d2",
                border_radius=12,
                elevation=4,
                on_click=lambda e: print("Extension clicked!")
            )
            page.add(ext)
        
        ft.run(main)
        ```
    """

    def __init__(
        self,
        src: Optional[str] = None,
        title: Optional[str] = None,
        subtitle: Optional[str] = None,
        background_color: Optional[str] = None,
        text_color: Optional[str] = None,
        border_radius: Optional[float] = None,
        border_width: Optional[float] = None,
        border_color: Optional[str] = None,
        padding: Optional[Union[int, float, ft.Padding]] = None,
        margin: Optional[Union[int, float, ft.Margin]] = None,
        animation_type: Optional[str] = None,
        animation_duration: Optional[Union[int, float, Duration]] = None,
        animation_curve: Optional[AnimationCurve] = None,
        clickable: Optional[bool] = None,
        elevation: Optional[float] = None,
        opacity: Optional[float] = None,
        on_click: Optional[Callable[[ft.ControlEvent], None]] = None,
        on_hover: Optional[Callable[[ft.ControlEvent], None]] = None,
        on_animation_complete: Optional[Callable[[ft.ControlEvent], None]] = None,
        **kwargs
    ):
        """
        Initialize the FletExtension control.
        
        Args:
            src: Primary content text to display
            title: Title text shown at the top
            subtitle: Subtitle text shown at the bottom
            background_color: Background color (hex, named, or CSS color)
            text_color: Text color for all text elements
            border_radius: Corner radius for rounded borders
            border_width: Width of the border line
            border_color: Color of the border
            padding: Internal spacing around content
            margin: External spacing around the control
            animation_type: Type of animation ('fade', 'scale', 'slide', 'rotate')
            animation_duration: Duration of animations in milliseconds
            animation_curve: Animation easing curve
            clickable: Whether the control responds to clicks
            elevation: Shadow elevation (0-24)
            opacity: Transparency level (0.0-1.0)
            on_click: Callback function for click events
            on_hover: Callback function for hover events
            on_animation_complete: Callback function when animation finishes
            **kwargs: Additional control properties
        """
        super().__init__(**kwargs)
        
        self.src = src
        self.title = title
        self.subtitle = subtitle
        self.background_color = background_color
        self.text_color = text_color
        self.border_radius = border_radius if border_radius is not None else 8.0
        self.border_width = border_width if border_width is not None else 0.0
        self.border_color = border_color
        self.padding = padding if padding is not None else 16.0
        self.margin = margin if margin is not None else 0.0
        self.animation_type = animation_type
        self.animation_duration = animation_duration if animation_duration is not None else 1.0
        self.animation_curve = animation_curve if animation_curve is not None else AnimationCurve.EASE_IN_OUT
        self.clickable = clickable if clickable is not None else True
        self.elevation = elevation if elevation is not None else 2.0
        self.opacity = opacity if opacity is not None else 1.0
        self.on_click = on_click
        self.on_hover = on_hover
        self.on_animation_complete = on_animation_complete
        
        # Initialize animation properties
        self.animation = None
        self.custom_animation = None
        self.font_size = 14.0

    # Content properties
    src: Optional[str] = field(default=None)
    """Primary content text to display in the extension.
    
    This is the main text content that will be prominently displayed.
    Supports plain text and can be updated dynamically.
    """
    
    title: Optional[str] = field(default=None)
    """Title text displayed at the top of the extension.
    
    Used for headings or primary labels. Typically rendered with
    larger font size and bold styling.
    """
    
    subtitle: Optional[str] = field(default=None)
    """Subtitle text displayed at the bottom of the extension.
    
    Used for secondary information, descriptions, or status text.
    Typically rendered with smaller font size.
    """
    
    # Visual properties
    background_color: Optional[str] = field(default=None)
    """Background color of the extension container.
    
    Accepts hex colors (#RRGGBB), named colors (red, blue), or CSS colors.
    Example: '#e3f2fd', 'lightblue', 'rgba(227, 242, 253, 0.8)'
    """
    
    text_color: Optional[str] = field(default=None)
    """Color of all text elements in the extension.
    
    Applied to title, src, and subtitle text. Accepts same formats
    as background_color. Defaults to theme text color if not specified.
    """
    
    border_color: Optional[str] = field(default=None)
    """Color of the border.
    
    Only visible when border_width > 0. Accepts same color formats
    as background_color.
    """
    
    border_width: Optional[float] = field(default=0.0)
    """Width of the border line in logical pixels.
    
    Set to 0 for no border. Typical values range from 1-4 pixels.
    Requires border_color to be visible.
    """
    
    border_radius: Optional[float] = field(default=8.0)
    """Corner radius for rounded borders in logical pixels.
    
    Controls how rounded the corners appear. Common values:
    - 0: Sharp corners
    - 4-8: Slightly rounded
    - 12-16: Moderately rounded
    - 20+: Very rounded
    """
    
    # Size and spacing
    font_size: Optional[float] = field(default=14.0)
    """Font size for text elements in logical pixels.
    
    Applied to the main src text. Title and subtitle may use
    proportionally larger/smaller sizes based on this value.
    """
    
    padding: Optional[float] = field(default=16.0)
    """Internal spacing around the content in logical pixels.
    
    Can be a single number for uniform padding, or ft.Padding object
    for different values per side.
    """
    
    margin: Optional[float] = field(default=0.0)
    """External spacing around the control in logical pixels.
    
    Can be a single number for uniform margin, or ft.Margin object
    for different values per side.
    """
    
    # Animation properties
    animation_duration: Optional[float] = field(default=1.0)
    """Duration of animations in seconds.
    
    Can be a float for seconds. Typical values: 0.2-0.5 for quick animations, 
    1.0-2.0 for slower effects. Used when triggering animations programmatically.
    """
    
    animation_type: Optional[str] = field(default=None)
    """Type of animation to apply when triggered.
    
    Supported animation types:
    - 'fade': Fade in/out effect
    - 'scale': Scale up/down effect
    - 'slide': Slide in/out effect
    - 'rotate': Rotation effect
    
    Use trigger_animation() or trigger_animation_async() to start animations.
    """
    
    animation: Optional[AnimationValue] = field(default=None)
    """Flet native animation configuration using AnimationValue.
    
    For advanced animation control. If specified, overrides
    animation_duration and animation_curve settings.
    """
    
    custom_animation: Optional[AnimationValue] = field(default=None)
    """Custom animation configuration using Flet's Animation class.
    
    Can be used alongside the main animation for layered effects.
    Allows fine-grained control over animation timing and curves.
    """
    
    animation_curve: Optional[AnimationCurve] = field(default=AnimationCurve.EASE_IN_OUT)
    """Animation easing curve for smooth transitions.
    
    Common curves:
    - EASE_IN_OUT: Smooth start and end (default)
    - LINEAR: Constant speed
    - EASE_IN: Slow start, fast end
    - EASE_OUT: Fast start, slow end
    - BOUNCE_IN/OUT: Bouncing effect
    """
    
    # Interactive properties
    clickable: Optional[bool] = field(default=True)
    """Whether the extension responds to click events.
    
    When True, the control will:
    - Show hover effects
    - Trigger on_click events when clicked
    - Display appropriate cursor on hover
    """
    
    elevation: Optional[float] = field(default=2.0)
    """Material Design shadow elevation (0-24).
    
    Controls the depth appearance of the control:
    - 0: No shadow (flat)
    - 1-4: Subtle shadow
    - 6-12: Moderate shadow
    - 16-24: Strong shadow (use sparingly)
    """
    
    opacity: Optional[float] = field(default=1.0)
    """Transparency level of the entire control (0.0 to 1.0).
    
    - 0.0: Completely transparent (invisible)
    - 0.5: Semi-transparent
    - 1.0: Completely opaque (default)
    
    Useful for fade effects or disabled states.
    """
    
    # Event handlers
    on_click: Optional[Callable[[ft.ControlEvent], None]] = field(default=None)
    """Callback function triggered when the control is clicked.
    
    Args:
        e (ft.ControlEvent): Event object containing click information
        
    Example:
        ```python
        def handle_click(e):
            print(f"Extension clicked! Control: {e.control}")
            e.control.src = "Clicked!"
            e.control.update()
            
        ext = FletExtension(on_click=handle_click)
        ```
    """
    
    on_hover: Optional[Callable[[ft.ControlEvent], None]] = field(default=None)
    """Callback function triggered when mouse enters/exits the control.
    
    Args:
        e (ft.ControlEvent): Event object with hover information
        
    Note:
        The event data contains information about hover state.
        Use e.data to determine if mouse entered ("true") or exited ("false").
        
    Example:
        ```python
        def handle_hover(e):
            if e.data == "true":
                e.control.elevation = 8
            else:
                e.control.elevation = 2
            e.control.update()
            
        ext = FletExtension(on_hover=handle_hover)
        ```
    """
    
    on_animation_complete: Optional[Callable[[ft.ControlEvent], None]] = field(default=None)
    """Callback function triggered when an animation finishes.
    
    Args:
        e (ft.ControlEvent): Event object containing animation completion info
        
    This is useful for chaining animations or updating UI state
    after an animation completes.
    
    Example:
        ```python
        def handle_animation_done(e):
            print("Animation completed!")
            e.control.src = "Animation finished"
            e.control.update()
            
        ext = FletExtension(on_animation_complete=handle_animation_done)
        ```
    """
    
    # Methods for triggering events from Python
    async def trigger_animation_async(self, animation_type: str = "fade", custom_animation: Optional[AnimationValue] = None):
        """
        Trigger an animation asynchronously on the Flutter side.
        
        This method communicates with the Flutter control to start an animation.
        The animation will use the control's current animation_duration and
        animation_curve settings.
        
        Args:
            animation_type: Type of animation to trigger
                - 'fade': Fade in/out effect
                - 'scale': Scale up/down effect  
                - 'slide': Slide in/out effect
                - 'rotate': Rotation effect
            custom_animation: Optional custom animation configuration
                
        Raises:
            ValueError: If animation_type is not supported
            
        Example:
            ```python
            # Trigger a scale animation
            await ext.trigger_animation_async("scale")
            
            # Trigger with custom duration
            ext.animation_duration = 0.5
            await ext.trigger_animation_async("fade")
            ```
        """
        valid_types = ["fade", "scale", "slide", "rotate"]
        if animation_type not in valid_types:
            raise ValueError(f"Invalid animation_type '{animation_type}'. Must be one of: {valid_types}")
            
        # Use provided custom_animation or create default
        if custom_animation is None:
            custom_animation = Animation(
                duration=Duration(milliseconds=int(self.animation_duration * 1000)),
                curve=self.animation_curve or AnimationCurve.EASE_IN_OUT
            )
        
        # Set the custom_animation property
        self.custom_animation = custom_animation
        
        try:
            return await self._invoke_method(
                "trigger_animation",
                {
                    "animation_type": animation_type
                }
            )
        except Exception as e:
            print(f"Debug: Failed to trigger animation '{animation_type}': {e}")
            raise
    
    async def update_content_async(self, src: str = None, title: str = None, subtitle: str = None):
        """
        Update content dynamically and asynchronously.
        
        This method updates the content properties and refreshes the control
        to reflect the changes immediately.
        
        Args:
            src: New main content text to display
            title: New title text (optional)
            subtitle: New subtitle text (optional)
            
        Example:
            ```python
            # Update all content
            await ext.update_content_async(
                src="New content!",
                title="Updated Title",
                subtitle="Updated subtitle"
            )
            
            # Update only main content
            await ext.update_content_async(src="Just new content")
            ```
        """
        # Update local properties if provided
        if src is not None:
            self.src = str(src)
        if title is not None:
            self.title = str(title)
        if subtitle is not None:
            self.subtitle = str(subtitle)
            
        try:
            return await self._invoke_method(
                "update_content",
                {
                    "src": src,
                    "title": title,
                    "subtitle": subtitle
                }
            )
        except Exception as e:
            print(f"Debug: Failed to update content: {e}")
            raise
    

    
    def trigger_animation(self, animation_type: str = "fade", custom_animation: Optional[AnimationValue] = None):
        """
        Trigger an animation synchronously on the Flutter side.
        
        This is the synchronous version of trigger_animation_async().
        Use this when you don't need to await the animation trigger.
        
        Args:
            animation_type: Type of animation to trigger
                - 'fade': Fade in/out effect
                - 'scale': Scale up/down effect
                - 'slide': Slide in/out effect
                - 'rotate': Rotation effect
            custom_animation: Optional custom animation configuration
                
        Raises:
            ValueError: If animation_type is not supported
            
        Example:
            ```python
            # Trigger animation in event handler
            def on_button_click(e):
                ext.trigger_animation("scale")
                
            # Trigger multiple animations
            ext.trigger_animation("fade")
            time.sleep(1)  # Wait for animation
            ext.trigger_animation("scale")
            ```
        """
        valid_types = ["fade", "scale", "slide", "rotate"]
        if animation_type not in valid_types:
            raise ValueError(f"Invalid animation_type '{animation_type}'. Must be one of: {valid_types}")
            
        # Use provided custom_animation or create default
        if custom_animation is None:
            custom_animation = Animation(
                duration=Duration(milliseconds=int(self.animation_duration * 1000)),
                curve=self.animation_curve or AnimationCurve.EASE_IN_OUT
            )
        
        # Set the custom_animation property
        self.custom_animation = custom_animation
        
        try:
            self._invoke_method(
                "trigger_animation",
                {
                    "animation_type": animation_type
                }
            )
        except Exception as e:
            print(f"Debug: Failed to trigger animation '{animation_type}': {e}")
            raise
    
    def update_content(self, src: str = None, title: str = None, subtitle: str = None):
        """
        Update content dynamically (sync version).
        
        This is the synchronous version of update_content_async().
        The control will be updated immediately to reflect the changes.
        
        Args:
            src: New main content text to display (optional)
            title: New title text (optional)
            subtitle: New subtitle text (optional)
            
        Example:
            ```python
            # Update all content in event handler
            def on_button_click(e):
                ext.update_content(
                    src="Button clicked!",
                    title="Updated",
                    subtitle="Status changed"
                )
                
            # Update only main content
            ext.update_content(src="New content only")
            ```
        """
        # Update local properties if provided
        if src is not None:
            self.src = str(src)
        if title is not None:
            self.title = str(title)
        if subtitle is not None:
            self.subtitle = str(subtitle)
            
        try:
            self._invoke_method(
                "update_content",
                {
                    "src": src,
                    "title": title,
                    "subtitle": subtitle
                }
            )
        except Exception as e:
            print(f"Debug: Failed to update content: {e}")
            raise

File: src/flutter/flet_extension/lib/src/flet_extension.dart
import 'package:flet/flet.dart';
import 'package:flutter/material.dart';
import 'dart:async';

/// A comprehensive Flet extension control demonstrating advanced Flutter integration.
/// 
/// This control serves as a template for creating custom Flet extensions with:
/// - Rich visual customization and Material Design support
/// - Multiple animation types with smooth transitions
/// - Interactive event handling and state management
/// - Professional error handling and debugging support
/// - Clean architecture suitable for developers of all levels
/// 
/// The control communicates with the Python side through method channels
/// and provides real-time updates for dynamic content changes.
class FletExtensionControl extends StatefulWidget {
  /// The Flet control configuration from Python.
  /// 
  /// Contains all properties, styling, and event handlers
  /// defined on the Python side.
  final Control control;

  /// Creates a FletExtension control.
  /// 
  /// The [control] parameter is required and contains all configuration
  /// from the Python side including styling, content, and event handlers.
  const FletExtensionControl({super.key, required this.control});

  @override
  State<FletExtensionControl> createState() => _FletExtensionControlState();
}

/// State management for the FletExtension control.
/// 
/// Handles animation controllers, user interactions, and communication
/// with the Python side through method channels. Provides smooth animations
/// and responsive UI updates while maintaining clean state management.
class _FletExtensionControlState extends State<FletExtensionControl>
    with TickerProviderStateMixin {
  // Animation system
  /// Main animation controller for all animation types.
  /// Duration and curve are configurable from Python side.
  late AnimationController _animationController;
  
  /// Fade animation for opacity transitions (1.0 to 0.0).
  late Animation<double> _fadeAnimation;
  
  /// Scale animation for size transitions (1.0 to 1.2).
  late Animation<double> _scaleAnimation;
  
  /// Slide animation for position transitions (zero to offset).
  late Animation<Offset> _slideAnimation;
  
  /// Rotation animation for 360-degree turns (0.0 to 1.0).
  late Animation<double> _rotateAnimation;
  
  // Interaction state
  /// Tracks whether the control is currently being hovered.
  bool _isHovered = false;
  
  /// Current active animation type ('fade', 'scale', 'slide', 'rotate', or 'none').
  String _currentAnimationType = 'none';

  @override
  void initState() {
    super.initState();
    debugPrint("Extension initState: ${widget.control.id}");
    
    try {
      // Initialize animation controller with default duration
      // Duration can be overridden by custom_animation property
      _animationController = AnimationController(
        duration: const Duration(milliseconds: 500),
        vsync: this,
      );
      
      // Add listener to update state during animation
      _animationController.addListener(() {
        setState(() {
          // This will trigger rebuilds during animation
        });
      });
      
      // Initialize all animation types with smooth default curves
      // These will be reconfigured when animations are triggered
      _initializeAnimations();
      
      // Set up animation completion monitoring
      _animationController.addStatusListener((status) {
        if (status == AnimationStatus.dismissed) {
          // Animation fully completed (including reverse)
          // Trigger completion event before resetting animation type
          if (_currentAnimationType != 'none') {
            _triggerAnimationComplete();
          }
          setState(() {
            _currentAnimationType = 'none';
          });
          debugPrint("Animation fully completed, reset to 'none'");
        } else if (status == AnimationStatus.completed) {
          debugPrint("Animation completed: $_currentAnimationType");
        }
      });
      
      // Register for method calls from Python side
      widget.control.addInvokeMethodListener(_invokeMethod);
      
      debugPrint("Extension initialization completed successfully");
    } catch (e) {
      debugPrint("Error during extension initialization: $e");
      // Continue with basic functionality even if some features fail
    }
  }
  
  /// Initializes all animation objects with default configurations.
  /// 
  /// These animations will be reconfigured with custom curves and durations
  /// when specific animations are triggered from the Python side.
  void _initializeAnimations() {
    const defaultCurve = Curves.easeInOut;
    
    _fadeAnimation = Tween<double>(begin: 1.0, end: 0.0).animate(
      CurvedAnimation(parent: _animationController, curve: defaultCurve),
    );
    
    _scaleAnimation = Tween<double>(begin: 1.0, end: 1.2).animate(
      CurvedAnimation(parent: _animationController, curve: defaultCurve),
    );
    
    _slideAnimation = Tween<Offset>(begin: Offset.zero, end: const Offset(0.1, 0)).animate(
      CurvedAnimation(parent: _animationController, curve: defaultCurve),
    );
    
    _rotateAnimation = Tween<double>(begin: 0.0, end: 1.0).animate(
      CurvedAnimation(parent: _animationController, curve: defaultCurve),
    );
  }

  @override
  void dispose() {
    debugPrint("Extension dispose: ${widget.control.id}");
    try {
      // Clean up animation controller to prevent memory leaks
      _animationController.dispose();
      debugPrint("Animation controller disposed successfully");
    } catch (e) {
      debugPrint("Error during animation controller disposal: $e");
    }
    widget.control.removeInvokeMethodListener(_invokeMethod);
    super.dispose();
  }
  
  /// Handles method calls from the Python side.
  /// 
  /// Supports the following methods:
  /// - `trigger_animation`: Starts an animation with the specified type
  /// - `update_content`: Updates control properties and triggers a rebuild
  /// 
  /// Returns a boolean indicating success, or throws an exception on error.
  Future<dynamic> _invokeMethod(String name, dynamic args) async {
    debugPrint("Extension method invoked: $name with args: $args");
    
    try {
      switch (name) {
        case "trigger_animation":
          final animationType = args?["animation_type"] as String? ?? "fade";
          debugPrint("Triggering animation: $animationType");
          return _triggerAnimation(animationType);
          
        case "update_content":
          debugPrint("Updating content with new properties");
          Map<String, dynamic>? typedArgs;
          if (args != null) {
            typedArgs = Map<String, dynamic>.from(args as Map);
          }
          return _updateContent(typedArgs);
          
        default:
          final errorMsg = "Unknown FletExtension method: $name";
          debugPrint(errorMsg);
          throw Exception(errorMsg);
      }
    } catch (e) {
      final errorMsg = "Error in FletExtension method $name: $e";
      debugPrint(errorMsg);
      rethrow;
    }
  }
  
  /// Retrieves the animation curve from custom_animation property.
  /// 
  /// Returns the curve specified in the custom_animation configuration,
  /// or defaults to [Curves.easeInOut] for smooth animations.
  Curve _getCurveFromAnimation() {
    try {
      final customAnimation = widget.control.getAnimation("custom_animation");
      return customAnimation?.curve ?? Curves.easeInOut;
    } catch (e) {
      debugPrint("Error getting animation curve: $e, using default");
      return Curves.easeInOut;
    }
  }

  /// Triggers an animation of the specified type.
  /// 
  /// Supported animation types:
  /// - 'fade': Opacity transition (visible to transparent and back)
  /// - 'scale': Size transition (normal to 1.2x scale and back)
  /// - 'slide': Position transition (normal to offset and back)
  /// - 'rotate': Rotation transition (0 to 360 degrees and back)
  /// 
  /// The animation uses custom duration and curve if specified in custom_animation,
  /// otherwise defaults to 500ms with easeInOut curve.
  /// 
  /// Returns true if animation started successfully, false otherwise.
  bool _triggerAnimation(String animationType) {
    debugPrint("Starting animation: $animationType");
    
    try {
      // Validate animation type
      const validTypes = ['fade', 'scale', 'slide', 'rotate'];
      if (!validTypes.contains(animationType)) {
        debugPrint("Warning: Unknown animation type '$animationType', using 'fade'");
        animationType = 'fade';
      }
      
      // Reset animation controller to initial state
      _animationController.reset();
      
      // Update current animation type
      setState(() {
        _currentAnimationType = animationType;
      });
      debugPrint("Current animation type set to: $_currentAnimationType");
      
      // Apply custom animation configuration if available
      final customAnimation = widget.control.getAnimation("custom_animation");
      if (customAnimation != null) {
        _animationController.duration = customAnimation.duration;
        debugPrint("Using custom animation duration: ${customAnimation.duration}");
      }
      
      // Get animation curve from configuration
      final animationCurve = _getCurveFromAnimation();
      
      // Reconfigure animations with current settings
      _fadeAnimation = Tween<double>(begin: 1.0, end: 0.0).animate(
        CurvedAnimation(parent: _animationController, curve: animationCurve),
      );
      
      _scaleAnimation = Tween<double>(begin: 1.0, end: 1.2).animate(
        CurvedAnimation(parent: _animationController, curve: animationCurve),
      );
      
      _slideAnimation = Tween<Offset>(begin: Offset.zero, end: const Offset(0.1, 0)).animate(
        CurvedAnimation(parent: _animationController, curve: animationCurve),
      );
      
      _rotateAnimation = Tween<double>(begin: 0.0, end: 1.0).animate(
        CurvedAnimation(parent: _animationController, curve: animationCurve),
      );
      
      // Execute animation sequence based on type
      switch (animationType) {
        case 'fade':
        case 'scale':
        case 'slide':
        case 'rotate':
          debugPrint("Starting forward animation for: $animationType");
          _animationController.forward().then((_) {
            debugPrint("Forward animation completed for: $animationType, starting reverse");
            _animationController.reverse().then((_) {
              debugPrint("Reverse animation completed for: $animationType");
              // Animation complete event will be triggered by status listener
            });
          });
          break;
        default:
          debugPrint("Starting single direction animation for: $animationType");
          _animationController.forward().then((_) {
            debugPrint("Single direction animation completed for: $animationType");
            // Animation complete event will be triggered by status listener
          });
      }
      
      return true;
    } catch (e) {
      debugPrint("Error triggering animation '$animationType': $e");
      return false;
    }
  }
  
  /// Updates the control's content and properties.
  /// 
  /// Takes a map of property updates from the Python side and applies them
  /// to the control, then triggers a rebuild to reflect the changes.
  /// 
  /// Returns true if the update was successful, false if args were null
  /// or an error occurred.
  bool _updateContent(Map<String, dynamic>? args) {
    if (args == null) {
      debugPrint("Update content called with null arguments");
      return false;
    }
    
    try {
      debugPrint("Updating control properties: ${args.keys.join(', ')}");
      
      // Update control properties with new values
      widget.control.update(args);
      
      // Trigger rebuild to reflect changes
      setState(() {});
      
      debugPrint("Content update completed successfully");
      return true;
    } catch (e) {
      debugPrint("Error updating content: $e");
      return false;
    }
  }

  /// Triggers the animation_complete event to notify the Python side.
  /// 
  /// Sends event data including the animation type and timestamp
  /// when an animation sequence finishes.
  void _triggerAnimationComplete() {
    try {
      // Capture the current animation type before it gets reset
      final completedAnimationType = _currentAnimationType;
      
      final eventData = {
        "animation_type": completedAnimationType,
        "timestamp": DateTime.now().millisecondsSinceEpoch,
      };
      
      widget.control.triggerEvent("animation_complete", eventData);
      debugPrint("Animation complete event triggered for: $completedAnimationType");
    } catch (e) {
      debugPrint("Error triggering animation complete event: $e");
    }
  }
  
  /// Handles click events and notifies the Python side.
  /// 
  /// Triggered when the user taps or clicks on the control.
  /// Sends event data including timestamp and click position.
  void _handleClick() {
    try {
      final eventData = {
        "timestamp": DateTime.now().millisecondsSinceEpoch,
        "position": "center",
      };
      
      widget.control.triggerEvent("click", eventData);
      debugPrint("Click event triggered");
    } catch (e) {
      debugPrint("Error handling click event: $e");
    }
  }
  
  /// Handles hover events and updates the visual state.
  /// 
  /// Updates the hover state and notifies the Python side when
  /// the mouse enters or exits the control area.
  void _handleHover(bool isHovered) {
    try {
      setState(() {
        _isHovered = isHovered;
      });
      
      final eventData = {
        "is_hovered": isHovered,
        "timestamp": DateTime.now().millisecondsSinceEpoch,
      };
      
      widget.control.triggerEvent("hover", eventData);
      debugPrint("Hover event triggered: $isHovered");
    } catch (e) {
      debugPrint("Error handling hover event: $e");
    }
  }

  /// Builds the widget tree for the FletExtension control.
  /// 
  /// Creates a rich, interactive widget with:
  /// - Customizable content (title, main text, subtitle)
  /// - Material Design styling (colors, borders, shadows)
  /// - Smooth animations (fade, scale, slide, rotate)
  /// - Interactive features (hover effects, click handling)
  /// 
  /// All properties are configurable from the Python side and updates
  /// are reflected immediately through setState calls.
  @override
  Widget build(BuildContext context) {
    debugPrint(
        "Extension build: ${widget.control.id} (${widget.control.hashCode})");
    
    try {
      // Extract content properties with safe defaults
      final src = widget.control.getString("src", "flet is good")!;
      final title = widget.control.getString("title");
      final subtitle = widget.control.getString("subtitle");
      
      // Extract visual styling properties
      final backgroundColor = widget.control.getString("background_color");
      final textColor = widget.control.getString("text_color");
      final borderColor = widget.control.getString("border_color");
      final borderWidth = widget.control.getDouble("border_width", 0.0)!;
      final borderRadius = widget.control.getDouble("border_radius", 8.0)!;
      final fontSize = widget.control.getDouble("font_size", 14.0)!;
      final padding = widget.control.getDouble("padding", 16.0)!;
      final elevation = widget.control.getDouble("elevation", 2.0)!;
      final opacity = widget.control.getDouble("opacity", 1.0)!;
      final clickable = widget.control.getBool("clickable", true)!;
      
      debugPrint("Building with animation type: $_currentAnimationType");
    
      // Build content widgets with proper styling
      final List<Widget> contentWidgets = [];
      
      // Add title if provided
      if (title != null && title.isNotEmpty) {
        contentWidgets.add(
          Text(
            title,
            style: TextStyle(
              fontSize: fontSize + 4,
              fontWeight: FontWeight.bold,
              color: textColor != null ? parseColor(textColor, Theme.of(context)) : null,
            ),
          ),
        );
        contentWidgets.add(const SizedBox(height: 8));
      }
      
      // Add main content (always present)
      contentWidgets.add(
        Text(
          src,
          style: TextStyle(
            fontSize: fontSize,
            color: textColor != null ? parseColor(textColor, Theme.of(context)) : null,
          ),
        ),
      );
      
      // Add subtitle if provided
      if (subtitle != null && subtitle.isNotEmpty) {
        contentWidgets.add(const SizedBox(height: 8));
        contentWidgets.add(
          Text(
            subtitle,
            style: TextStyle(
              fontSize: fontSize - 2,
              fontStyle: FontStyle.italic,
              color: textColor != null 
                  ? parseColor(textColor, Theme.of(context)) 
                  : Colors.grey[600],
            ),
          ),
        );
      }
      
      // Create content column
      Widget content = Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        mainAxisSize: MainAxisSize.min,
        children: contentWidgets,
      );
    
      // Apply current animation type to content
      Widget animatedContent = content;
      switch (_currentAnimationType) {
        case 'fade':
          animatedContent = FadeTransition(
            opacity: _fadeAnimation,
            child: content,
          );
          break;
        case 'scale':
          animatedContent = ScaleTransition(
            scale: _scaleAnimation,
            child: content,
          );
          break;
        case 'slide':
          animatedContent = SlideTransition(
            position: _slideAnimation,
            child: content,
          );
          break;
        case 'rotate':
          animatedContent = RotationTransition(
            turns: _rotateAnimation,
            child: content,
          );
          break;
        default:
          // No animation active, use content as is
          animatedContent = content;
          break;
      }
    
      // Create styled container with Material Design elements
      Widget styledContent = Container(
        padding: EdgeInsets.all(padding),
        decoration: BoxDecoration(
          color: parseColor(backgroundColor, Theme.of(context)),
          border: borderWidth > 0 ? Border.all(
            color: parseColor(borderColor, Theme.of(context)) ?? Colors.grey,
            width: borderWidth,
          ) : null,
          borderRadius: BorderRadius.circular(borderRadius),
          boxShadow: elevation > 0 ? [
            BoxShadow(
              color: Colors.black.withAlpha(25),
              blurRadius: elevation,
              offset: Offset(0, elevation / 2),
            ),
          ] : null,
        ),
        child: animatedContent,
      );
      
      // Apply opacity if less than fully opaque
      if (opacity < 1.0) {
        styledContent = Opacity(
          opacity: opacity.clamp(0.0, 1.0), // Ensure valid opacity range
          child: styledContent,
        );
      }
      
      // Add interactive features if enabled
      if (clickable) {
        styledContent = MouseRegion(
          onEnter: (_) => _handleHover(true),
          onExit: (_) => _handleHover(false),
          child: GestureDetector(
            onTap: _handleClick,
            child: AnimatedContainer(
              duration: const Duration(milliseconds: 200),
              transform: Matrix4.identity()..scale(_isHovered ? 1.02 : 1.0),
              child: styledContent,
            ),
          ),
        );
      }
      
      return ConstrainedControl(
        control: widget.control,
        child: styledContent,
      );
      
    } catch (e) {
      debugPrint("Error building FletExtension widget: $e");
      
      // Return a fallback widget in case of errors
      return ConstrainedControl(
        control: widget.control,
        child: Container(
          padding: const EdgeInsets.all(16.0),
          decoration: BoxDecoration(
            color: Colors.red[50],
            border: Border.all(color: Colors.red, width: 1),
            borderRadius: BorderRadius.circular(8),
          ),
          child: const Text(
            "Error: Unable to render FletExtension",
            style: TextStyle(color: Colors.red),
          ),
        ),
      );
    }
  }
}
